src/core/internal/traits.d

```diff
@@ -258,14 +258,63 @@ template hasElaborateCopyConstructor(S)
     }
     else static if (is(S == struct))
     {
-        enum hasElaborateCopyConstructor = __traits(hasMember, S, "__xpostblit");
+        enum hasElaborateCopyConstructor = __traits(hasMember, S, "__xpostblit") ||
+             () {
+                static if (__traits(hasMember, S, "__ctor"))
+                {
+                    static foreach (f; __traits(getOverloads, S, "__ctor"))
+                    {{
+                        bool r = __traits(compiles, {
+                                auto p = &f;
+
+                                // Check if ctor is callable with lval or rval
+                                S s;
+                                (*p)(s);
+
+                                // Check that ctor is't callable with rval
+                                static assert (!__traits(compiles, (*p)(S()) ));
+                        });
+                        if (r) return true;
+                    }}
+                }
+                return false;
+             }();
     }
     else
     {
         enum bool hasElaborateCopyConstructor = false;
     }
 }
 
+@safe unittest
+{
+    static struct S
+    {
+        int x;
+        this(return scope ref typeof(this) rhs) { }
+        this(int x, int y) {}
+    }
+
+    static assert(hasElaborateCopyConstructor!S);
```

Comment:
Making `S` non-`static` will make the assert fail, and I can't figure out why. It's unclear to me if the behaviour is ok.